File: scenarios/advanced/mobile/android-compose/RichTextEditor.kt
package com.example.accessibilitybenchmark

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RichTextEditor() {
    var content by remember { mutableStateOf("") }
    var isBold by remember { mutableStateOf(false) }
    var isItalic by remember { mutableStateOf(false) }
    var isUnderline by remember { mutableStateOf(false) }
    var fontSize by remember { mutableStateOf(16f) }
    var textColor by remember { mutableStateOf(Color.Black) }
    var backgroundColor by remember { mutableStateOf(Color.White) }
    var alignment by remember { mutableStateOf(TextAlign.Start) }
    var showWordCount by remember { mutableStateOf(false) }
    var isUndoAvailable by remember { mutableStateOf(false) }
    var isRedoAvailable by remember { mutableStateOf(false) }
    var history by remember { mutableStateOf(listOf("")) }
    var historyIndex by remember { mutableStateOf(0) }
    var showFontSizePicker by remember { mutableStateOf(false) }
    var showColorPicker by remember { mutableStateOf(false) }
    var showInsertMenu by remember { mutableStateOf(false) }
    var colorPickerType by remember { mutableStateOf(ColorPickerType.TEXT) }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Rich Text Editor") },
                actions = {
                    TextButton(onClick = { /* Save */ }) {
                        Text("Save")
                    }
                    TextButton(onClick = { /* Export */ }) {
                        Text("Export")
                    }
                    TextButton(onClick = { showWordCount = !showWordCount }) {
                        Text("Stats")
                    }
                }
            )
        }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
        ) {
            // Toolbar
            Toolbar(
                isBold = isBold,
                isItalic = isItalic,
                isUnderline = isUnderline,
                fontSize = fontSize,
                textColor = textColor,
                backgroundColor = backgroundColor,
                alignment = alignment,
                isUndoAvailable = isUndoAvailable,
                isRedoAvailable = isRedoAvailable,
                onBoldClick = { isBold = !isBold },
                onItalicClick = { isItalic = !isItalic },
                onUnderlineClick = { isUnderline = !isUnderline },
                onFontSizeClick = { showFontSizePicker = true },
                onTextColorClick = { 
                    colorPickerType = ColorPickerType.TEXT
                    showColorPicker = true 
                },
                onBackgroundColorClick = { 
                    colorPickerType = ColorPickerType.BACKGROUND
                    showColorPicker = true 
                },
                onAlignmentClick = { alignment = it },
                onUndoClick = { 
                    if (historyIndex > 0) {
                        historyIndex--
                        content = history[historyIndex]
                        isUndoAvailable = historyIndex > 0
                        isRedoAvailable = true
                    }
                },
                onRedoClick = { 
                    if (historyIndex < history.size - 1) {
                        historyIndex++
                        content = history[historyIndex]
                        isUndoAvailable = true
                        isRedoAvailable = historyIndex < history.size - 1
                    }
                },
                onInsertClick = { showInsertMenu = true }
            )

            // Word Count Panel
            if (showWordCount) {
                WordCountPanel(
                    wordCount = content.trim().split("\\s+".toRegex()).filter { it.isNotEmpty() }.size,
                    charCount = content.length
                )
            }

            // Editor
            Editor(
                content = content,
                onContentChange = { newContent ->
                    content = newContent
                    val newHistory = history.take(historyIndex + 1) + newContent
                    history = newHistory
                    historyIndex = newHistory.size - 1
                    isUndoAvailable = historyIndex > 0
                    isRedoAvailable = historyIndex < newHistory.size - 1
                },
                isBold = isBold,
                isItalic = isItalic,
                isUnderline = isUnderline,
                fontSize = fontSize,
                textColor = textColor,
                backgroundColor = backgroundColor,
                alignment = alignment
            )
        }
    }

    // Font Size Picker
    if (showFontSizePicker) {
        FontSizePicker(
            selectedSize = fontSize,
            onSizeSelected = { size ->
                fontSize = size
                showFontSizePicker = false
            },
            onDismiss = { showFontSizePicker = false }
        )
    }

    // Color Picker
    if (showColorPicker) {
        ColorPicker(
            selectedColor = if (colorPickerType == ColorPickerType.TEXT) textColor else backgroundColor,
            colorType = colorPickerType,
            onColorSelected = { color ->
                if (colorPickerType == ColorPickerType.TEXT) {
                    textColor = color
                } else {
                    backgroundColor = color
                }
                showColorPicker = false
            },
            onDismiss = { showColorPicker = false }
        )
    }

    // Insert Menu
    if (showInsertMenu) {
        InsertMenu(
            onInsert = { type ->
                when (type) {
                    InsertType.IMAGE -> content += "\n[Image placeholder]\n"
                    InsertType.LINK -> content += "\n[Link placeholder]\n"
                    InsertType.TABLE -> content += "\n[Table placeholder]\n"
                    InsertType.LIST -> content += "\n• List item\n"
                }
                showInsertMenu = false
            },
            onDismiss = { showInsertMenu = false }
        )
    }
}

@Composable
fun Toolbar(
    isBold: Boolean,
    isItalic: Boolean,
    isUnderline: Boolean,
    fontSize: Float,
    textColor: Color,
    backgroundColor: Color,
    alignment: TextAlign,
    isUndoAvailable: Boolean,
    isRedoAvailable: Boolean,
    onBoldClick: () -> Unit,
    onItalicClick: () -> Unit,
    onUnderlineClick: () -> Unit,
    onFontSizeClick: () -> Unit,
    onTextColorClick: () -> Unit,
    onBackgroundColorClick: () -> Unit,
    onAlignmentClick: (TextAlign) -> Unit,
    onUndoClick: () -> Unit,
    onRedoClick: () -> Unit,
    onInsertClick: () -> Unit
) {
    LazyRow(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .padding(vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        contentPadding = PaddingValues(horizontal = 16.dp)
    ) {
        // Undo/Redo Group
        item {
            ToolbarButton(
                icon = Icons.Default.Undo,
                enabled = isUndoAvailable,
                onClick = onUndoClick
            )
        }
        item {
            ToolbarButton(
                icon = Icons.Default.Redo,
                enabled = isRedoAvailable,
                onClick = onRedoClick
            )
        }
        item { ToolbarSeparator() }

        // Formatting Group
        item {
            ToolbarButton(
                icon = Icons.Default.FormatBold,
                selected = isBold,
                onClick = onBoldClick
            )
        }
        item {
            ToolbarButton(
                icon = Icons.Default.FormatItalic,
                selected = isItalic,
                onClick = onItalicClick
            )
        }
        item {
            ToolbarButton(
                icon = Icons.Default.FormatUnderlined,
                selected = isUnderline,
                onClick = onUnderlineClick
            )
        }
        item { ToolbarSeparator() }

        // Font Size Group
        item {
            ToolbarButton(
                text = "${fontSize.toInt()}pt",
                onClick = onFontSizeClick
            )
        }
        item {
            ColorButton(
                color = textColor,
                onClick = onTextColorClick
            )
        }
        item {
            ColorButton(
                color = backgroundColor,
                onClick = onBackgroundColorClick
            )
        }
        item { ToolbarSeparator() }

        // Alignment Group
        item {
            ToolbarButton(
                icon = Icons.Default.FormatAlignLeft,
                selected = alignment == TextAlign.Start,
                onClick = { onAlignmentClick(TextAlign.Start) }
            )
        }
        item {
            ToolbarButton(
                icon = Icons.Default.FormatAlignCenter,
                selected = alignment == TextAlign.Center,
                onClick = { onAlignmentClick(TextAlign.Center) }
            )
        }
        item {
            ToolbarButton(
                icon = Icons.Default.FormatAlignRight,
                selected = alignment == TextAlign.End,
                onClick = { onAlignmentClick(TextAlign.End) }
            )
        }
        item { ToolbarSeparator() }

        // Insert Group
        item {
            ToolbarButton(
                icon = Icons.Default.Add,
                onClick = onInsertClick
            )
        }
    }
}

@Composable
fun ToolbarButton(
    icon: androidx.compose.ui.graphics.vector.ImageVector? = null,
    text: String? = null,
    selected: Boolean = false,
    enabled: Boolean = true,
    onClick: () -> Unit
) {
    Box(
        modifier = Modifier
            .size(40.dp)
            .background(
                if (selected) MaterialTheme.colorScheme.primary.copy(alpha = 0.2f)
                else Color.Transparent,
                RoundedCornerShape(6.dp)
            )
            .clickable(enabled = enabled) { onClick() },
        contentAlignment = Alignment.Center
    ) {
        if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = if (enabled) MaterialTheme.colorScheme.onSurface else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f)
            )
        } else if (text != null) {
            Text(
                text = text,
                style = MaterialTheme.typography.labelMedium,
                color = if (enabled) MaterialTheme.colorScheme.onSurface else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f)
            )
        }
    }
}

@Composable
fun ColorButton(
    color: Color,
    onClick: () -> Unit
) {
    Box(
        modifier = Modifier
            .size(20.dp)
            .background(color, CircleShape)
            .border(1.dp, MaterialTheme.colorScheme.outline, CircleShape)
            .clickable { onClick() }
    )
}

@Composable
fun ToolbarSeparator() {
    Box(
        modifier = Modifier
            .width(1.dp)
            .height(32.dp)
            .background(MaterialTheme.colorScheme.outline)
    )
}

@Composable
fun WordCountPanel(
    wordCount: Int,
    charCount: Int
) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .padding(vertical = 8.dp),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = "Words: $wordCount | Characters: $charCount",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
fun Editor(
    content: String,
    onContentChange: (String) -> Unit,
    isBold: Boolean,
    isItalic: Boolean,
    isUnderline: Boolean,
    fontSize: Float,
    textColor: Color,
    backgroundColor: Color,
    alignment: TextAlign
) {
    BasicTextField(
        value = content,
        onValueChange = onContentChange,
        modifier = Modifier
            .fillMaxSize()
            .background(backgroundColor)
            .padding(16.dp),
        textStyle = TextStyle(
            fontSize = fontSize.sp,
            color = textColor,
            fontWeight = if (isBold) FontWeight.Bold else FontWeight.Normal,
            fontStyle = if (isItalic) FontStyle.Italic else FontStyle.Normal,
            textDecoration = if (isUnderline) TextDecoration.Underline else TextDecoration.None,
            textAlign = alignment
        ),
        cursorBrush = SolidColor(MaterialTheme.colorScheme.primary),
        keyboardOptions = KeyboardOptions(imeAction = ImeAction.Default),
        keyboardActions = KeyboardActions()
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FontSizePicker(
    selectedSize: Float,
    onSizeSelected: (Float) -> Unit,
    onDismiss: () -> Unit
) {
    val fontSizes = listOf(12f, 14f, 16f, 18f, 20f, 24f, 28f, 32f, 36f, 48f)

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Font Size") },
        text = {
            LazyColumn {
                items(fontSizes) { size ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable { onSizeSelected(size) }
                            .padding(vertical = 8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(
                            text = "${size.toInt()}pt",
                            style = MaterialTheme.typography.bodyLarge.copy(fontSize = size.sp),
                            modifier = Modifier.weight(1f)
                        )
                        if (selectedSize == size) {
                            Icon(
                                Icons.Default.Check,
                                contentDescription = null,
                                tint = MaterialTheme.colorScheme.primary
                            )
                        }
                    }
                }
            }
        },
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text("Done")
            }
        }
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ColorPicker(
    selectedColor: Color,
    colorType: ColorPickerType,
    onColorSelected: (Color) -> Unit,
    onDismiss: () -> Unit
) {
    val colors = listOf(
        Color.Black to "Black",
        Color.Gray to "Gray",
        Color.Red to "Red",
        Color.Orange to "Orange",
        Color.Yellow to "Yellow",
        Color.Green to "Green",
        Color.Blue to "Blue",
        Color.Magenta to "Purple",
        Color.Cyan to "Cyan"
    )

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(if (colorType == ColorPickerType.TEXT) "Text Color" else "Background Color") },
        text = {
            LazyColumn(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                items(colors.chunked(5)) { rowColors ->
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        rowColors.forEach { (color, name) ->
                            Box(
                                modifier = Modifier
                                    .weight(1f)
                                    .aspectRatio(1f)
                                    .background(color, CircleShape)
                                    .border(
                                        2.dp,
                                        if (selectedColor == color) MaterialTheme.colorScheme.primary else Color.Transparent,
                                        CircleShape
                                    )
                                    .clickable { onColorSelected(color) }
                            )
                        }
                    }
                }
            }
        },
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text("Done")
            }
        }
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun InsertMenu(
    onInsert: (InsertType) -> Unit,
    onDismiss: () -> Unit
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Insert") },
        text = {
            LazyColumn(
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                items(InsertType.values()) { type ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable { onInsert(type) }
                            .padding(vertical = 8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Icon(
                            imageVector = when (type) {
                                InsertType.IMAGE -> Icons.Default.Image
                                InsertType.LINK -> Icons.Default.Link
                                InsertType.TABLE -> Icons.Default.TableChart
                                InsertType.LIST -> Icons.Default.List
                            },
                            contentDescription = null,
                            modifier = Modifier.size(32.dp)
                        )
                        Spacer(modifier = Modifier.width(16.dp))
                        Text(
                            text = type.name.lowercase().replaceFirstChar { it.uppercase() },
                            style = MaterialTheme.typography.bodyLarge
                        )
                    }
                }
            }
        },
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text("Done")
            }
        }
    )
}

enum class ColorPickerType {
    TEXT, BACKGROUND
}

enum class InsertType {
    IMAGE, LINK, TABLE, LIST
}

File: frameworks/mobile/flutter/lib/rich_text_editor.dart
import 'package:flutter/material.dart';

class RichTextEditor extends StatefulWidget {
  const RichTextEditor({Key? key}) : super(key: key);

  @override
  State<RichTextEditor> createState() => _RichTextEditorState();
}

class _RichTextEditorState extends State<RichTextEditor> {
  final TextEditingController _controller = TextEditingController();
  bool isBold = false;
  bool isItalic = false;
  bool isUnderline = false;
  double fontSize = 16;
  Color textColor = Colors.black;
  Color backgroundColor = Colors.white;
  TextAlign alignment = TextAlign.left;
  bool showStylePanel = false;
  bool showMediaPanel = false;
  bool showColorPicker = false;
  String colorType = 'text'; // text or background
  int wordCount = 0;
  int charCount = 0;
  bool isUndoAvailable = false;
  bool isRedoAvailable = false;
  List<String> history = [''];
  int historyIndex = 0;
  bool showWordCount = false;
  bool showInsertMenu = false;

  final List<double> fontSizes = [12, 14, 16, 18, 20, 24, 28, 32, 36, 48];
  final List<Color> colors = [
    Colors.black, Colors.grey[800]!, Colors.grey[600]!, Colors.grey[400]!, Colors.grey[200]!,
    Colors.red, Colors.orange, Colors.amber, Colors.green, Colors.blue,
    Colors.purple, Colors.pink, Colors.white, Colors.red[100]!, Colors.green[100]!,
  ];

  @override
  void initState() {
    super.initState();
    _controller.addListener(_updateCounts);
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  void _updateCounts() {
    final text = _controller.text;
    setState(() {
      wordCount = text.trim().split(RegExp(r'\s+')).where((word) => word.isNotEmpty).length;
      charCount = text.length;
    });
  }

  void _updateHistory() {
    final newHistory = history.sublist(0, historyIndex + 1);
    newHistory.add(_controller.text);
    setState(() {
      history = newHistory;
      historyIndex = newHistory.length - 1;
      isUndoAvailable = historyIndex > 0;
      isRedoAvailable = historyIndex < newHistory.length - 1;
    });
  }

  void _handleUndo() {
    if (historyIndex > 0) {
      setState(() {
        historyIndex--;
        _controller.text = history[historyIndex];
        isUndoAvailable = historyIndex > 0;
        isRedoAvailable = true;
      });
    }
  }

  void _handleRedo() {
    if (historyIndex < history.length - 1) {
      setState(() {
        historyIndex++;
        _controller.text = history[historyIndex];
        isUndoAvailable = true;
        isRedoAvailable = historyIndex < history.length - 1;
      });
    }
  }

  void _handleBold() {
    setState(() {
      isBold = !isBold;
    });
    // In a real implementation, this would apply bold formatting to selected text
  }

  void _handleItalic() {
    setState(() {
      isItalic = !isItalic;
    });
    // In a real implementation, this would apply italic formatting to selected text
  }

  void _handleUnderline() {
    setState(() {
      isUnderline = !isUnderline;
    });
    // In a real implementation, this would apply underline formatting to selected text
  }

  void _handleFontSizeChange(double size) {
    setState(() {
      fontSize = size;
      showStylePanel = false;
    });
  }

  void _handleColorChange(Color color) {
    setState(() {
      if (colorType == 'text') {
        textColor = color;
      } else {
        backgroundColor = color;
      }
      showColorPicker = false;
    });
  }

  void _handleAlignmentChange(TextAlign align) {
    setState(() {
      alignment = align;
    });
  }

  void _handleInsertImage() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Image insertion functionality would be implemented here')),
    );
    setShowInsertMenu(false);
  }

  void _handleInsertLink() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Link insertion functionality would be implemented here')),
    );
    setShowInsertMenu(false);
  }

  void _handleInsertTable() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Table insertion functionality would be implemented here')),
    );
    setShowInsertMenu(false);
  }

  void _handleInsertList() {
    final listItem = '• List item\n';
    _controller.text += listItem;
    _updateHistory();
    setShowInsertMenu(false);
  }

  void _handleSave() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Document saved successfully')),
    );
  }

  void _handleExport() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Export'),
        content: const Text('Choose export format'),
        actions: [
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ScaffoldMessenger.of(context).showSnackBar(
                const SnackBar(content: Text('Exporting as PDF...')),
              );
            },
            child: const Text('PDF'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ScaffoldMessenger.of(context).showSnackBar(
                const SnackBar(content: Text('Exporting as Word...')),
              );
            },
            child: const Text('Word'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              ScaffoldMessenger.of(context).showSnackBar(
                const SnackBar(content: Text('Exporting as HTML...')),
              );
            },
            child: const Text('HTML'),
          ),
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: const Text('Cancel'),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey[100],
      appBar: AppBar(
        title: const Text('Rich Text Editor'),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
        actions: [
          IconButton(
            icon: const Icon(Icons.save),
            onPressed: _handleSave,
          ),
          IconButton(
            icon: const Icon(Icons.download),
            onPressed: _handleExport,
          ),
          IconButton(
            icon: const Icon(Icons.analytics),
            onPressed: () => setState(() => showWordCount = !showWordCount),
          ),
        ],
      ),
      body: Column(
        children: [
          _buildToolbar(),
          if (showWordCount) _buildWordCountPanel(),
          Expanded(
            child: _buildEditor(),
          ),
        ],
      ),
    );
  }

  Widget _buildToolbar() {
    return Container(
      padding: const EdgeInsets.symmetric(vertical: 8),
      color: Colors.white,
      child: SingleChildScrollView(
        scrollDirection: Axis.horizontal,
        child: Row(
          children: [
            _buildToolbarGroup([
              _buildToolbarButton(
                icon: Icons.undo,
                isActive: isUndoAvailable,
                onPressed: isUndoAvailable ? _handleUndo : null,
              ),
              _buildToolbarButton(
                icon: Icons.redo,
                isActive: isRedoAvailable,
                onPressed: isRedoAvailable ? _handleRedo : null,
              ),
            ]),
            _buildToolbarSeparator(),
            _buildToolbarGroup([
              _buildToolbarButton(
                icon: Icons.format_bold,
                isActive: isBold,
                onPressed: _handleBold,
              ),
              _buildToolbarButton(
                icon: Icons.format_italic,
                isActive: isItalic,
                onPressed: _handleItalic,
              ),
              _buildToolbarButton(
                icon: Icons.format_underlined,
                isActive: isUnderline,
                onPressed: _handleUnderline,
              ),
            ]),
            _buildToolbarSeparator(),
            _buildToolbarGroup([
              _buildToolbarButton(
                text: fontSize.toInt().toString(),
                onPressed: () => setState(() => showStylePanel = true),
              ),
              _buildColorButton(
                color: textColor,
                onPressed: () {
                  setState(() {
                    colorType = 'text';
                    showColorPicker = true;
                  });
                },
              ),
              _buildColorButton(
                color: backgroundColor,
                onPressed: () {
                  setState(() {
                    colorType = 'background';
                    showColorPicker = true;
                  });
                },
              ),
            ]),
            _buildToolbarSeparator(),
            _buildToolbarGroup([
              _buildToolbarButton(
                icon: Icons.format_align_left,
                isActive: alignment == TextAlign.left,
                onPressed: () => _handleAlignmentChange(TextAlign.left),
              ),
              _buildToolbarButton(
                icon: Icons.format_align_center,
                isActive: alignment == TextAlign.center,
                onPressed: () => _handleAlignmentChange(TextAlign.center),
              ),
              _buildToolbarButton(
                icon: Icons.format_align_right,
                isActive: alignment == TextAlign.right,
                onPressed: () => _handleAlignmentChange(TextAlign.right),
              ),
            ]),
            _buildToolbarSeparator(),
            _buildToolbarGroup([
              _buildToolbarButton(
                icon: Icons.add,
                onPressed: () => setState(() => showInsertMenu = true),
              ),
              _buildToolbarButton(
                icon: Icons.analytics,
                onPressed: () => setState(() => showWordCount = !showWordCount),
              ),
            ]),
          ],
        ),
      ),
    );
  }

  Widget _buildToolbarGroup(List<Widget> children) {
    return Row(
      children: children.map((child) => Padding(
        padding: const EdgeInsets.symmetric(horizontal: 4),
        child: child,
      )).toList(),
    );
  }

  Widget _buildToolbarSeparator() {
    return Container(
      width: 1,
      height: 32,
      color: Colors.grey[300],
      margin: const EdgeInsets.symmetric(horizontal: 8),
    );
  }

  Widget _buildToolbarButton({
    IconData? icon,
    String? text,
    bool isActive = false,
    VoidCallback? onPressed,
  }) {
    return GestureDetector(
      onTap: onPressed,
      child: Container(
        width: 40,
        height: 40,
        decoration: BoxDecoration(
          color: isActive ? Colors.blue[100] : Colors.transparent,
          borderRadius: BorderRadius.circular(6),
        ),
        child: Center(
          child: icon != null
              ? Icon(
                  icon,
                  size: 20,
                  color: isActive ? Colors.blue : Colors.grey[600],
                )
              : Text(
                  text ?? '',
                  style: TextStyle(
                    fontSize: 14,
                    fontWeight: FontWeight.w600,
                    color: Colors.grey[600],
                  ),
                ),
        ),
      ),
    );
  }

  Widget _buildColorButton({
    required Color color,
    required VoidCallback onPressed,
  }) {
    return GestureDetector(
      onTap: onPressed,
      child: Container(
        width: 20,
        height: 20,
        decoration: BoxDecoration(
          color: color,
          borderRadius: BorderRadius.circular(10),
          border: Border.all(color: Colors.grey[300]!),
        ),
      ),
    );
  }

  Widget _buildWordCountPanel() {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      color: Colors.grey[200],
      child: Text(
        'Words: $wordCount | Characters: $charCount',
        style: const TextStyle(fontSize: 12, color: Colors.grey),
        textAlign: TextAlign.center,
      ),
    );
  }

  Widget _buildEditor() {
    return Container(
      margin: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: backgroundColor,
        borderRadius: BorderRadius.circular(8),
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.1),
            blurRadius: 4,
            offset: const Offset(0, 2),
          ),
        ],
      ),
      child: TextField(
        controller: _controller,
        style: TextStyle(
          fontSize: fontSize,
          color: textColor,
          fontWeight: isBold ? FontWeight.bold : FontWeight.normal,
          fontStyle: isItalic ? FontStyle.italic : FontStyle.normal,
          decoration: isUnderline ? TextDecoration.underline : TextDecoration.none,
        ),
        textAlign: alignment,
        maxLines: null,
        expands: true,
        decoration: const InputDecoration(
          hintText: 'Start typing your document...',
          hintStyle: TextStyle(color: Colors.grey),
          border: InputBorder.none,
          contentPadding: EdgeInsets.all(16),
        ),
        onChanged: (text) => _updateHistory(),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey[100],
      appBar: AppBar(
        title: const Text('Rich Text Editor'),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
        actions: [
          IconButton(
            icon: const Icon(Icons.save),
            onPressed: _handleSave,
          ),
          IconButton(
            icon: const Icon(Icons.download),
            onPressed: _handleExport,
          ),
          IconButton(
            icon: const Icon(Icons.analytics),
            onPressed: () => setState(() => showWordCount = !showWordCount),
          ),
        ],
      ),
      body: Column(
        children: [
          _buildToolbar(),
          if (showWordCount) _buildWordCountPanel(),
          Expanded(
            child: _buildEditor(),
          ),
        ],
      ),
      bottomSheet: showStylePanel ? _buildStylePanel() : null,
      bottomSheet: showColorPicker ? _buildColorPicker() : null,
      bottomSheet: showInsertMenu ? _buildInsertMenu() : null,
    );
  }

  Widget _buildStylePanel() {
    return Container(
      height: 300,
      decoration: const BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.vertical(top: Radius.circular(20)),
      ),
      child: Column(
        children: [
          Container(
            padding: const EdgeInsets.all(16),
            decoration: const BoxDecoration(
              border: Border(bottom: BorderSide(color: Colors.grey, width: 0.5)),
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                const Text(
                  'Font Size',
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
                ),
                IconButton(
                  onPressed: () => setState(() => showStylePanel = false),
                  icon: const Icon(Icons.close),
                ),
              ],
            ),
          ),
          Expanded(
            child: ListView.builder(
              itemCount: fontSizes.length,
              itemBuilder: (context, index) {
                final size = fontSizes[index];
                final isSelected = fontSize == size;
                return ListTile(
                  title: Text(
                    '${size.toInt()}pt',
                    style: TextStyle(fontSize: size),
                  ),
                  trailing: isSelected ? const Icon(Icons.check, color: Colors.blue) : null,
                  onTap: () => _handleFontSizeChange(size),
                );
              },
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildColorPicker() {
    return Container(
      height: 300,
      decoration: const BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.vertical(top: Radius.circular(20)),
      ),
      child: Column(
        children: [
          Container(
            padding: const EdgeInsets.all(16),
            decoration: const BoxDecoration(
              border: Border(bottom: BorderSide(color: Colors.grey, width: 0.5)),
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  colorType == 'text' ? 'Text Color' : 'Background Color',
                  style: const TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
                ),
                IconButton(
                  onPressed: () => setState(() => showColorPicker = false),
                  icon: const Icon(Icons.close),
                ),
              ],
            ),
          ),
          Expanded(
            child: GridView.builder(
              padding: const EdgeInsets.all(16),
              gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
                crossAxisCount: 5,
                crossAxisSpacing: 8,
                mainAxisSpacing: 8,
              ),
              itemCount: colors.length,
              itemBuilder: (context, index) {
                final color = colors[index];
                final isSelected = (colorType == 'text' ? textColor : backgroundColor) == color;
                return GestureDetector(
                  onTap: () => _handleColorChange(color),
                  child: Container(
                    decoration: BoxDecoration(
                      color: color,
                      borderRadius: BorderRadius.circular(20),
                      border: Border.all(
                        color: isSelected ? Colors.blue : Colors.grey[300]!,
                        width: isSelected ? 3 : 1,
                      ),
                    ),
                    child: isSelected
                        ? const Icon(
                            Icons.check,
                            color: Colors.white,
                            size: 16,
                          )
                        : null,
                  ),
                );
              },
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildInsertMenu() {
    return Container(
      height: 200,
      decoration: const BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.vertical(top: Radius.circular(20)),
      ),
      child: Column(
        children: [
          Container(
            padding: const EdgeInsets.all(16),
            decoration: const BoxDecoration(
              border: Border(bottom: BorderSide(color: Colors.grey, width: 0.5)),
            ),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                const Text(
                  'Insert',
                  style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
                ),
                IconButton(
                  onPressed: () => setState(() => showInsertMenu = false),
                  icon: const Icon(Icons.close),
                ),
              ],
            ),
          ),
          Expanded(
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: [
                _buildInsertMenuItem(
                  icon: Icons.image,
                  label: 'Image',
                  onTap: _handleInsertImage,
                ),
                _buildInsertMenuItem(
                  icon: Icons.link,
                  label: 'Link',
                  onTap: _handleInsertLink,
                ),
                _buildInsertMenuItem(
                  icon: Icons.grid_on,
                  label: 'Table',
                  onTap: _handleInsertTable,
                ),
                _buildInsertMenuItem(
                  icon: Icons.list,
                  label: 'List',
                  onTap: _handleInsertList,
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildInsertMenuItem({
    required IconData icon,
    required String label,
    required VoidCallback onTap,
  }) {
    return GestureDetector(
      onTap: onTap,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(icon, size: 32, color: Colors.blue),
          const SizedBox(height: 8),
          Text(
            label,
            style: const TextStyle(fontSize: 12, color: Colors.grey),
          ),
        ],
      ),
    );
  }
}
